### ISSUE TITLE
Discrepancy Between OpenCost and Monitoring Values in GKE

### ISSUE THREAD

**Issue by RCarva** [opencost/opencost#2777]
I'm using GKE, and when I make a request to the OpenCost API aggregating by namespaces:

```
curl -G http://localhost:9003/allocation \                                                                                                                                                                           
  -d window=6h \
  -d aggregate=namespace \
  -d resolution=1m > input.json
```

Then, I receive some namespaces with values equal to 0. This error persists when I use different aggregate values, as shown in the example below:
```
"window": {
                    "start": "2024-06-06T15:26:54Z",
                    "end": "2024-06-06T21:26:54Z"
                },
                "start": "2024-06-06T15:27:00Z",
                "end": "2024-06-06T21:26:00Z",
                "minutes": 359,
                "cpuCores": 0,
                "cpuCoreRequestAverage": 0,
                "cpuCoreUsageAverage": 0,
                "cpuCoreHours": 0,
                "cpuCost": 0,
                "cpuCostAdjustment": 0,
                "cpuEfficiency": 0,
                "gpuCount": 0,
                "gpuHours": 0,
                "gpuCost": 0,
                "gpuCostAdjustment": 0,
                "networkTransferBytes": 0,
                "networkReceiveBytes": 0,
                "networkCost": 0,
                "networkCrossZoneCost": 0,
                "networkCrossRegionCost": 0,
                "networkInternetCost": 0,
                "networkCostAdjustment": 0,
                "loadBalancerCost": 0,
                "loadBalancerCostAdjustment": 0,
                "pvBytes": 0,
                "pvByteHours": 0,
                "pvCost": 0,
                "pvs": null,
                "pvCostAdjustment": 0,
                "ramBytes": 0,
                "ramByteRequestAverage": 0,
                "ramByteUsageAverage": 0,
                "ramByteHours": 0,
                "ramCost": 0,
                "ramCostAdjustment": 0,
                "ramEfficiency": 0,
                "externalCost": 0,
                "sharedCost": 0,
                "totalCost": 0,
                "totalEfficiency": 0,
                "lbAllocations": null
            },
```
Some namespaces are correct, but there are 5 of them with the same error. However, in the Google Cloud Monitoring, I see that the values are different from 0. In the case above, the real value of "cpuCoreRequestAverage" should be 0.08. I would like to understand what could be causing this discrepancy. Any insights or suggestions would be greatly appreciated!

note: it recognizes all the pods/containers and the window which the pod was running.

**Comment by mattray** [comment]
@RCarva which version of OpenCost are you using? Are you seeing this issue? https://github.com/opencost/opencost/issues/2742

**Comment by RCarva** [comment]
I'm using the version 1.109, and about the issue https://github.com/opencost/opencost/issues/2742, i received this logs in my opencost:

```
ERR Failed to lookup reserved instance data: googleapi: Error 400 │
│ opencost Details:                                                                                         │
│ opencost [                                                                                                │
│ opencost   {                                                                                              │
│ opencost     "@type": "type.googleapis.com/google.rpc.ErrorInfo",                                         │
│ opencost     "domain": "googleapis.com",                                                                  │
│ opencost     "metadatas": {                                                                               │
│ opencost       "method": "compute.v1.RegionCommitmentsService.AggregatedList",                            │
│ opencost       "service": "compute.googleapis.com"                                                        │
│ opencost     },                                                                                           │
│ opencost     "reason": "RESOURCE_PROJECT_INVALID"                                                         │
│ opencost   }                                                                                              │
│ opencost ]                                                                                                │
│ opencost , invalidParameter 
```
But only that error in the logs, and some "inf".

**Comment by mattray** [comment]
I assume you have a valid API key?

**Comment by RCarva** [comment]
Yeah, my API key has access to all Google APIs, it was created 1 year ago.

**Comment by charalamm** [comment]
I am seeing the same behavior in azure when aggregating by label. I have not checked all our labels but all 10 I checked had a lower `cpuCoreRequestAverage` than the one requested. We are using version 1.108.0

**Comment by mattray** [comment]
Could you please upgrade to 1.111.0 and report back?

**Comment by RafikFarhad** [comment]
I am using 1.111.0 and the error is still there:
```
Defaulted container "opencost" out of: opencost, opencost-ui
2024-06-29T06:51:32.18776874Z ??? Log level set to error
2024-06-29T06:51:32.818205367Z ERR Failed to lookup reserved instance data: googleapi: Error 400: Invalid resource field value in the request.
Details:
[
  {
    "@type": "type.googleapis.com/google.rpc.ErrorInfo",
    "domain": "googleapis.com",
    "metadatas": {
      "method": "compute.v1.RegionCommitmentsService.AggregatedList",
      "service": "compute.googleapis.com"
    },
    "reason": "RESOURCE_PROJECT_INVALID"
  }
]
, invalidParameter
```

My ServiceAccount and the BigQuery dataset are in different GCP projects. Could it be an issue? 
My sample config is like this:
```
{
  "gcp": {
    "bigQuery": [
      {
        "projectID": "project-1",
        "dataset": "billing_export_data",
        "table": "gcp_billing_export_resource_v1_aaa_bbb_ccc",
        "authorizer": {
          "authorizerType": "GCPServiceAccountKey",
          "key": {
            "type": "service_account",
            "project_id": "project-2",
            "private_key_id": "xxxxxxxxxxxx",
            "private_key": "",
            "client_email": "opencost@project-2.iam.gserviceaccount.com",
            "client_id": "xxxxxxxx",
            "auth_uri": "https://accounts.google.com/o/oauth2/auth",
            "token_uri": "https://oauth2.googleapis.com/token",
            "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
            "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/opencost%40project-2.iam.gserviceaccount.com",
            "universe_domain": "googleapis.com"
          }
        }
      }
    ]
  }
}
```
```

**Comment by mattray** [comment]
Just to level set you have 
```
opencost:
  exporter:
    cloudProviderApiKey: "BLAHBLAHBLAH-rB2Z1FnUakzLngd2poHU"
```
in your values.yaml that you're using with the Helm chart? I have a similar issue where I get `"ACCESS_TOKEN_SCOPE_INSUFFICIENT"` with my GKE but it still work(ish).

**Comment by RafikFarhad** [comment]
Thanks for looking into this.
Yes, I have tried both of the following formats:
```
opencost:
  exporter:
    cloudProviderApiKey: "xxxxx_yyyyyyyyyyyyy"
```
and
```
opencost:
  metrics:
  exporter:
    env:
      - name: CLOUD_PROVIDER_API_KEY
        valueFrom:
          secretKeyRef:
            name: <custom-secret-name>
            key: GcpApiKey
```
but no luck. My API key has `Cloud Billing API` permission so I am not expecting `ACCESS_TOKEN_SCOPE_INSUFFICIENT`.

**Comment by mattray** [comment]
Yeah, mine has that too but I'm in an environment that is trying to lock everything down. Could you humor me and create a new API key with everything just for testing?

**Comment by lmello** [comment]
After looking at the code, I am suspecting the issue lies in you having the user in a different project. 

To access a dataset in a project  (project-1) thats different from the project where the service account exists (project-2), the bigquery client need to use [DatasetInProject](https://pkg.go.dev/cloud.google.com/go/bigquery#Client.DatasetInProject) method. Didn't find any entry in opencost code for that method. Maybe it was implemented in a different way.

Any implementation would require opencost to get the project-id from the authorizer and check if it is the same project id or different than the project id defined for the integration.  Which I didn't find those types of checks anywhere in the [gcp](https://github.com/opencost/opencost/tree/develop/pkg/cloud/gcp) code. 

This looks like a  bug in opencost, as a work around, you could create a service account with access to billing in the SAME project where the billing data is.  

I am going to flag this as bug.

**Comment by mattray** [comment]
Seems like something Kubecost would run into.

**Comment by RafikFarhad** [comment]
FYI: I tried using a service account from the same project where the big query dataset is. Still facing this error.

**Comment by AjayTripathy** [comment]
cc @cliffcolvin FYI

**Comment by RafikFarhad** [comment]
This line is overriding the `projectID` [here](https://github.com/opencost/opencost/blob/805b231751843b995e6c56748d9b2d6abf7420bf/pkg/cloud/gcp/provider.go#L1094) that is being used to call the ` computeService.RegionCommitments.AggregatedList(gcp.ProjectID)` [here](https://github.com/opencost/opencost/blob/805b231751843b995e6c56748d9b2d6abf7420bf/pkg/cloud/gcp/provider.go#L1377).
Can anyone please verify? I can raise a PR with some guidance on the appropriate fix.

**Comment by cliffcolvin** [comment]
> This line is overriding the `projectID` [here](https://github.com/opencost/opencost/blob/805b231751843b995e6c56748d9b2d6abf7420bf/pkg/cloud/gcp/provider.go#L1094) that is being used to call the ` computeService.RegionCommitments.AggregatedList(gcp.ProjectID)` [here](https://github.com/opencost/opencost/blob/805b231751843b995e6c56748d9b2d6abf7420bf/pkg/cloud/gcp/provider.go#L1377). Can anyone please verify? I can raise a PR with some guidance on the appropriate fix.

@RafikFarhad Ive asked Sean to take a look here. We’ll get back with you when he has a chance to look.

**Comment by Sean-Holcomb** [comment]
@RafikFarhad I would say that it is overriding it some much as it is setting it but yes that is correct. That being said I am not seeing a good general purpose way to set that value in the OpenCost via helm chart, one fix here would to be to replace it with `ClusterProjectID` which is being set based of node labels to be the ProjectID where the cluster exists
```
computeService.RegionCommitments.AggregatedList(gcp.ClusterProjectID)
```

The configuration you posted previously for BigQuery exists in a different code path and isn't (and shouldn't be) used here.

 To be a bit more targeted here, are you seeing the issue that the original post is complaining about with the zeroed out values on certain aggregations, or are you just concerned because you are also seeing this log?

**Comment by pvnstty** [comment]
> @RafikFarhad I would say that it is overriding it some much as it is setting it but yes that is correct. That being said I am not seeing a good general purpose way to set that value in the OpenCost via helm chart, one fix here would to be to replace it with `ClusterProjectID` which is being set based of node labels to be the ProjectID where the cluster exists
> 
> ```
> computeService.RegionCommitments.AggregatedList(gcp.ClusterProjectID)
> ```
> 
> The configuration you posted previously for BigQuery exists in a different code path and isn't (and shouldn't be) used here.
> 
> To be a bit more targeted here, are you seeing the issue that the original post is complaining about with the zeroed out values on certain aggregations, or are you just concerned because you are also seeing this log?

I am also getting the same error. And my service account and the GKE cluster are in p1 account and the bigquery dataset is in p2 project. I created a SA in p1 with compute.viewer and jobuser roles and gave the data viewer role for the SA for dataset and table in p2. And also created an API key in p2 and enabled cloud billing api .  I get the kubernetes costs but for cloud costs,, the UI is showing - "There are no Cloud Cost integrations currently configured." So if you use ClusterProjectID, in this case again it will take the SA project id so not sure how it will fix the issue.  Can you please advise how we can fix this issue? Would you advice to try using the same project for everything?

**Comment by RafikFarhad** [comment]
@Sean-Holcomb 
> are you seeing the issue that the original post is complaining about with the zeroed out values on certain aggregations, or are you just concerned because you are also seeing this log?

Initially, I was hunting another error and thought this specific log is the issue. Then, I fixed this specific project id fix on my branch and this log is not there anymore, but my original issue is still there and making progress on debugging that (it is something to do with promql query to Grafana Cloud - still investigating slowly).

It would be appreciated if we solve this project id anomaly as it has the opportunity to make things confusing and not working for the scenario where the project id of SA and deployment do not match.  


@pvnstty I am not sure, I have the same as you described and I also tried with having everything on the same project. I ended up building my own image with this diff, that solved the `"reason": "RESOURCE_PROJECT_INVALID"` error specifically: 

(sorry for proving an image of the diff, I can not create a fork and provide a pr/branch with this)
![image](https://github.com/user-attachments/assets/2cae86a1-bc41-4c4e-9a46-f9f81c8c5ac7)

**Comment by pvnstty** [comment]
> @Sean-Holcomb
> 
> > are you seeing the issue that the original post is complaining about with the zeroed out values on certain aggregations, or are you just concerned because you are also seeing this log?
> 
> Initially, I was hunting another error and thought this specific log is the issue. Then, I fixed this specific project id fix on my branch and this log is not there anymore, but my original issue is still there and making progress on debugging that (it is something to do with promql query to Grafana Cloud - still investigating slowly).
> 
> It would be appreciated if we solve this project id anomaly as it has the opportunity to make things confusing and not working for the scenario where the project id of SA and deployment do not match.
> 
> @pvnstty I am not sure, I have the same as you described and I also tried with having everything on the same project. I ended up building my own image with this diff, that solved the `"reason": "RESOURCE_PROJECT_INVALID"` error specifically:
> 
> (sorry for proving an image of the diff, I can not create a fork and provide a pr/branch with this) ![image](https://private-user-images.githubusercontent.com/15923187/393638741-2cae86a1-bc41-4c4e-9a46-f9f81c8c5ac7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM3NDM0NDYsIm5iZiI6MTczMzc0MzE0NiwicGF0aCI6Ii8xNTkyMzE4Ny8zOTM2Mzg3NDEtMmNhZTg2YTEtYmM0MS00YzRlLTlhNDYtZjlmODFjOGM1YWM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA5VDExMTkwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU5ZjE4ZDZlNDM1MDFjNGVkZTNmOTg4OThjNjA4YjI0N2ZiMDg5ZDYwOGVlOTAxNGI1YTAwOGY3YmVhNjZmYWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IFaGkqsadsRq9_srGaWKg_lxY6iodAJZTG7E0nDT4_0)

@RafikFarhad  Thanks for the response. I am also using the same account for all the resources and still getting that error but I think as everyone said it shouldn't happen if they apply the fix. After troubleshooting , I found out that my problem was with the memory allotted. Now I increased the memory requests/limits and it is taking around 4GiB , now it seems working. But I am still testing as sometimes I am getting the same screen if we don't enable the cloud costs. Please check the container restarts.

**Comment by Sean-Holcomb** [comment]
@RafikFarhad It sounds like you are requesting a manual override for the project id variable via the helm chart and you are actually experiencing the issue that is outlined by the original poster.
